karapace/kafka_rest_apis/__init__.py

```diff
@@ -316,19 +322,14 @@ def init_admin_client(self):
                 self.log.exception("Unable to start admin client, retrying")
                 time.sleep(1)
 
-    async def close_producers(self):
-        if not self.producer_refs:
-            return
-        for prod in self.producer_refs:
-            self.log.info("Disposing of async producers")
-            await prod.stop()
-        self.producer_refs = None
-        self.producer_queue = None
-        return
-
     async def close(self) -> None:
         await super().close()
-        await self.close_producers()
+
+        async with self._async_producer_lock:
+            if self._async_producer is not None:
+                self.log.info("Disposing async producer")
+                await self._async_producer.stop()
```

Comment:
This needs `self._async_producer = None` after `.stop()`

Reply:
I thought about it, the problem is that the code doesn't seem to have a notion of services being restartable or use once.

I don't see why this should be made restartable, so there is no need to return it to the state of initialization (if we do so, another producer could be created while the server is shutting down). There is no real benefit in setting it to None and it may be surprising on some corner cases.